### PR TITLE
feat(frontend): migrate to typescript

### PR DESCRIPTION
**Pull request by zaxutic**
migrates the frontend from pure js to typescript

- modifies the API request helper function to also deserialise JSON responses if one is expected
    - this was done to make it easier to type the other api functions with the expected return type from the backend (Response is not generic)
    - request takes a generic parameter type T for the response type. if T is void (default), then it expects `{ jsonResp: false }` in params and vice versa
    - will only try to deserialise json if `jsonResp` is true (it dies if we try to otherwise)
    - throw error if http status not ok - create custom FetchError type with status, statusText, and the response object itself (if we need for whatever reason) as attributes
- fixes random type errors throughout the codebase that we didn't catch before
- removes `uuid` (we were using this to generate temporary ids in the frontend only (didn't get sent to backend), but this generates strings and all of our ids are expected to be numbers, making everything inconsistent and potentially break)
- adds eslint plugins/rules for typescript, and added some extra useful rules/plugins (and fixed the unexpectedly high number of linting errors)
- remove prop-types 🦀
- some things were super jank/broken and i had to fix/refactor them to make typescript happy
- remove deprecated html attributes (bruh) in favour of proper css

### CODE REVIEW

File: frontend/.eslintrc.json [hunk 1]
```diff
@@ -4,26 +4,59 @@
     "es2021": true
   },
   "extends": [
+    "plugin:@typescript-eslint/recommended",
+    "plugin:@typescript-eslint/recommended-requiring-type-checking",
     "plugin:react/recommended",
     "airbnb",
```
Comment (gtangelo) [inline frontend/.eslintrc.json:10]: why not also add `airbnb-typescript` 👀

**Review by gtangelo — APPROVED** [review]
I'm trusting your typescript changes to not break prod 😎

**Review by stapletm — APPROVED** [review]
INCREDIBLE

File: frontend/.eslintrc.json (hunk 1 above)
Comment (zaxutic) [reply frontend/.eslintrc.json:10]: true i should 👀

**Pull request merged**